runtime/lua/vim/lsp/util.lua

```diff
@@ -474,7 +474,20 @@ function M.apply_text_edits(text_edits, bufnr, offset_encoding)
       end
       has_eol_text_edit = true
     end
-    api.nvim_buf_set_text(bufnr, e.start_row, e.start_col, e.end_row, e.end_col, e.text)
+
+    local last_line = get_line(bufnr, e.end_row)
+
+    -- If the replacement is over full lines it is safer to use api.nvim_buf_set_lines
+    if e.end_col >= #last_line and e.start_col == 0 then
+      -- When replacing whole lines there seems to always be an extra newline at the end of the new
+      -- text. So we remove it.
+      if e.text[#e.text] == '' then
+        table.remove(e.text, #e.text)
```

Comment:
The test failure may be caused by `e.text` becoming empty here.

Reply:
Thanks!  That seems to have been the issue. I'm still not super happy about the whole solution but the two use cases I got errors for works with the current change. It would be awesome if someone could help me write a test for this.